### PR TITLE
[Wasm-GC] Fix some interactions between subtyping and recursion

### PR DESCRIPTION
**Pull request by takikawa**
#### 0d872184b6e17f15c34fe5f5c8ddd50a3269a129
<pre>
[Wasm-GC] Fix some interactions between subtyping and recursion
<a href="https://bugs.webkit.org/show_bug.cgi?id=247536">https://bugs.webkit.org/show_bug.cgi?id=247536</a>

Reviewed by Justin Michaud.

Interactions between recursive types and subtyping can be tricky,
because a `sub` clause can refer to another member of the same recursion
group as a supertype. How this supertype is dereferenced can affect
whether subtyping/type equality holds correctly.

This patch makes a few fixes to the representation of subtypes to
accommodate recursion. The supertype index of a Subtype should be a
recursive placeholder when pointing to another member of the same
recursion group (i.e., it should resolve to the projection). And the
structural subtype check needs to be deferred until after projections
are constructed.

In addition, reference subtype checking via displays is abstracted out
of the Subtype and put into a new RTT data structure. Every type in the
type section has a canonical RTT. This RTT is looked up to check
subtyping between references. Splitting out the RTT provides a better
logical separation between subtype declaration and checking, and makes
Subtype construction cleaner (as RTT creation can happen after recursive
references can be fully resolved).

The RTT can be used in the future to implement subtype-aware runtime
casts.

* JSTests/wasm/gc-spec-tests/type-subtyping.wast.js:
* JSTests/wasm/gc/sub.js:
(testSubDeclaration):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isSubtypeIndex):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseType):
(JSC::Wasm::SectionParser::parseRecursionGroup):
(JSC::Wasm::SectionParser::checkSubtypeValidity):
(JSC::Wasm::SectionParser::parseSubtype):
* Source/JavaScriptCore/wasm/WasmSectionParser.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::Projection::dump const):
(JSC::Wasm::TypeDefinition::tryCreateSubtype):
(JSC::Wasm::substituteParent):
(JSC::Wasm::TypeDefinition::replacePlaceholders const):
(JSC::Wasm::TypeDefinition::hasRecursiveReference const):
(JSC::Wasm::RTT::tryCreateRTT):
(JSC::Wasm::RTT::isSubRTT const):
(JSC::Wasm::SubtypeParameterTypes::translate):
(JSC::Wasm::TypeInformation::registerCanonicalRTTForType):
(JSC::Wasm::TypeInformation::canonicalRTTForType):
(JSC::Wasm::TypeInformation::tryCleanup):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::RecursionGroup::getType):
(JSC::Wasm::Subtype::Subtype):
(JSC::Wasm::Subtype::underlyingType const):
(JSC::Wasm::Subtype::getUnderlyingType):
(JSC::Wasm::RTT::RTT):
(JSC::Wasm::RTT::displaySize const):
(JSC::Wasm::RTT::displayEntry const):
(JSC::Wasm::RTT::setDisplayEntry):
(JSC::Wasm::RTT::allocatedRTTSize):
(JSC::Wasm::RTT::payload):
(JSC::Wasm::TypeDefinition::allocatedSubtypeSize):
(JSC::Wasm::Subtype::displayType const): Deleted.
(JSC::Wasm::Subtype::displaySize const): Deleted.
(JSC::Wasm::Subtype::getDisplayType): Deleted.

Canonical link: <a href="https://commits.webkit.org/257945@main">https://commits.webkit.org/257945@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/5191c2aa89a4bfbe19d9035f11de1ab2edbb084d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100416 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9578 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109731 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169990 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10489 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/150 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92819 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107608 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106195 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7938 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91208 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34587 "Passed tests") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22598 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77545 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90959 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3316 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24117 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/86975 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/774 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3306 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29118 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9428 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43607 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/89857 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5444 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5123 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20089 "Passed tests") | 
| | | | |

### CODE REVIEW

**Comment by takikawa** [comment]
I originally thought this was ready for review, but I've since found a bug in the approach. Here's a Wasm module that demonstrates this regression:

```
(module
  (rec (type $f1 (func))
       (type (struct))
       (type $s1 (sub $f1 (func))))
  (rec (type $f2 (func))
       (type $s2 (sub $f2 (func))))

  (func (param (ref null $f1)))
  (func (call 0 (ref.null $s2))) ;; $s2 is not a valid subtype of $f1 as they're in different recursion groups
)
```

The above module shouldn't validate, and doesn't in the Wasm spec interpreter. However, it will validate with this PR. The issue is that the recursion group's identity matters for checking equality of types in a subtype hierarchy, and that isn't accounted for here.

In any case, I've converted the PR back to a draft and will fix this issue.

File: Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp [hunk 1]
```diff
@@ -160,7 +160,10 @@ void Projection::dump(PrintStream& out) const
 {
     out.print("(");
     CommaPrinter comma;
-    TypeInformation::get(recursionGroup()).dump(out);
+    if (isPlaceholder())
+        out.print("<current-rec-group>");
```
Comment (takikawa) [inline Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:164]: Note: I'm hoping to more comprehensively improve type printing (for bug https://bugs.webkit.org/show_bug.cgi?id=247746) later. This change was the minimal change needed to make the tests work for now.

**Review by justinmichaud — APPROVED** [review]
r=me

**Comment by takikawa** [comment]
The only failing check on the last run was a windows-build issue which has gone away in the latest run, so I'll set merge-queue now.

**Pull request merged**